pkg/controller/cluster/cluster_controller_test.go

```diff
@@ -0,0 +1,980 @@
+	return "update"
+}
+
+func (ea expectedClusterStatusUpdateAction) validate(t *testing.T, action clientgotesting.Action) bool {
+	if action.GetSubresource() != "status" {
+		return false
+	}
+	updateAction, ok := action.(clientgotesting.UpdateAction)
+	if !ok {
+		t.Errorf("update action is not an UpdateAction: %t", updateAction)
+		return false
+	}
+	updatedObject := updateAction.GetObject()
+	cluster, ok := updatedObject.(*clusteroperator.Cluster)
+	if !ok {
+		t.Errorf("cluster status update action object is not a Cluster: %t", cluster)
+		return true
+	}
+	if e, a := ea.masterNodeGroups, cluster.Status.MasterNodeGroups; e != a {
+		t.Errorf("unexpected masterNodeGroups in cluster update status: expected %v, got %v", e, a)
+	}
+	if e, a := ea.computeNodeGroups, cluster.Status.ComputeNodeGroups; e != a {
+		t.Errorf("unexpected computeNodeGroups in cluster update status: expected %v, got %v", e, a)
+	}
+	return true
+}
+
+// validateControllerExpectations validates that the specified cluster is
+// expecting the specified number of creations and deletions of node groups.
+func validateControllerExpectations(t *testing.T, ctrlr *ClusterController, cluster *clusteroperator.Cluster, expectedAdds, expectedDeletes int) {
+	expectations, ok, err := ctrlr.expectations.GetExpectations(getKey(cluster, t))
+	switch {
+	case err != nil:
+		t.Errorf("error getting expecations: %v", cluster.GetName(), err)
+	case !ok:
+		if expectedAdds != 0 || expectedDeletes != 0 {
+			t.Errorf("no expectations found: expectedAdds %v, expectedDeletes %v", expectedAdds, expectedDeletes)
+		}
+	default:
+		actualsAdds, actualDeletes := expectations.GetExpectations()
+		if e, a := int64(expectedAdds), actualsAdds; e != a {
+			t.Errorf("unexpected number of adds in expectations: expected %v, got %v", e, a)
+		}
+		if e, a := int64(expectedDeletes), actualDeletes; e != a {
+			t.Errorf("unexpected number of deletes in expectations: expected %v, got %v", e, a)
+		}
+	}
+}
+
+// TestSyncClusterSteadyState tests syncing a cluster when it is in a steady
+// state.
+func TestSyncClusterSteadyState(t *testing.T) {
+	cases := []struct {
+		name     string
+		computes []string
+	}{
+		{
+			name: "only master",
+		},
+		{
+			name:     "single compute",
+			computes: []string{"compute1"},
+		},
+		{
+			name:     "multiple computes",
+			computes: []string{"compute1", "compute2"},
+		},
+	}
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			controller, clusterStore, nodeGroupStore, _, clusterOperatorClient := newTestClusterController()
+
+			cluster := newCluster(tc.computes...)
+			cluster.Status.MasterNodeGroups = 1
+			cluster.Status.ComputeNodeGroups = len(tc.computes)
+			clusterStore.Add(cluster)
+			newNodeGroups(nodeGroupStore, cluster, true, tc.computes...)
+
+			controller.syncCluster(getKey(cluster, t))
+
+			validateClientActions(t, clusterOperatorClient)
+
+			validateControllerExpectations(t, controller, cluster, 0, 0)
```

Comment:
Will be really nice to have this expectations framework example to work off, I'll be referencing back here once this goes in.